### ISSUE TITLE
[topic] Theming Web Components w/ Global CSS

### ISSUE THREAD

**Issue by thescientist13** [webcomponents-cg/docs-and-guides#21]
### Code of conduct

- [X] I agree to follow this project's code of conduct.

## Proposal
One of the great features introduced with Web Components was Shadow DOM.  However, this also introduced some real challenges with (most) active and current CSS styling libraries and tools, which have only ever had to assume / support a single "global" scope for CSS, and by association, all the massive amounts of projects and sites using them.

Examples include
- [Bootstrap](https://getbootstrap.com/)
- [Font Awesome](https://fontawesome.com/)
- [Tailwinds](https://tailwindcss.com/)
- [Normalize](http://necolas.github.io/normalize.css/)

## Description
While on the surface it might seem counter-intuitive to advocate for this when using Web Components, but I think it's important to remember that WCs are an umbrella of specifications, and so AFAIK, there is no rule that says you have to use _all_ features together.  In fact, I think one could argue that the "purity" of Shadow DOM could be pointed to as a friction point for adopting WCs because if you're using any of those existing CSS solutions and just want to componentize your app as custom elements, your barrier to entry just increased because of some of the only known work arounds you have to implement (as I know them anyway) are to
1.  Shim / inline that global CSS into the Shadow DOM of _every_ custom element, which will lead to significantly larger bundle sizes.
1.  Or, you can just turn Shadow DOM off.  This does mean that any 3rd party WCs you use need to support this too, or else they will introduce an impenetrable "shadow boundary".

I think ultimately this can help adoptees make better and more informed decisions and hopefully custom elements (with optional / value added Shadow DOM) can be that gateway into the wider world of WCs.

## Outline
1. Introduction of CSS and Shadow DOM and the central "conflict"
1. Option A (no Shadow DOM) w/ caveats - no `<slot>`, 3rd party shadow boundaries
1. Option B (inline all shared CSS) w/ caveats - bundle size bloat, tooling
1. Other option?
1. Transitioning from global CSS to Shadow DOM based CSS (aka _is_ there a happy path for this?)
1. Conclusion, link to spec and community channels, companion GitHub repo with examples, etc

In general I think the over-arching theme should be seen less as not advocating for Shadow DOM or "enabling" global CSS for ever, but rather, as I think is how the real world is, that _incremental_ adoption strategies should be possible.

I think it would be great to highlight that a team, developer, or large organization's design system can still take advantage of custom elements (an arguably higher ROI) even if they don't immediately also adopt and thus re-write their entire design system to satisfy Shadow DOM.  It should be very possible to do one, then the other as I think an all or nothing mentality does more harm than good.  (IMO).

## Links / References
Collection of relevant links, github issues, etc
- https://github.com/WICG/webcomponents/issues/909
- https://github.com/ProjectEvergreen/greenwood/discussions/763#discussioncomment-1477811
- https://github.com/fernandopasik/lit-redux-router/issues/60
- https://lit.dev/docs/components/shadow-dom/#implementing-createrenderroot

----

I think an article like this plays very well into one of the big themes / topics from our WC3 report, which was on design systems and theming w/ WCs.  This article could certainly be a stepping stone to that larger discussion combined with more articles that could hopefully provide a great guide on how to start a design system _today_.  This would benefit from sourcing a lot of community examples so hopefully we can get some solid community involvement here to really flesh out abstract with the tangible!   🙌 

### Are you interested in authoring this content?

- [X] Yes

**Comment by sashafirsov** [comment]
> Shim / inline that global CSS into the Shadow DOM of every custom element, which will lead to significantly larger bundle sizes.
That would not be needed if the resource could be referenced from within the page, say by ID. In current concept the resources reuse is underdeveloped. We could preload it via LINK, but can not apply by referencing that link declaratively, instead URI is the main link point. Once style would permit something like 
```css
@import ref("#my-css-or-link-tag-id")
```
the volume concern should go away along with bundle concept. Reused CSS sub-bundle in similar to JS fashion would be assembled by build toolchain and reused across WC. 

Of course the scope and base url for module has to be solved, which is part of css/html module discussion.

**Comment by sashafirsov** [comment]
> Disable Shadow DOM. This does mean that any 3rd party WCs you use need to support this too, or else they will introduce an impenetrable "shadow boundary".

It would be true if we apply to WC `mode=open` as it is of now. Most likely this mode would be kept for compatibility reasons, hence another mode something like `mode=inject` would reflect the WC acting almost as HTML include. Of course with templates/slots respected as in [light-dom-element](https://www.npmjs.com/package/light-dom-element). `mode=light` meant for customizable by container CSS.

mode= `closed`, `open`, `light`, or `inject` could co-exist happily. Even more, developer could make them act identical if the CSS could be reused as in previous comment. There are samples of [light](https://unpkg.com/light-dom-element@0.0.7/demo/ananke/content/en/index.html) and [shadow](https://unpkg.com/shadow-dom-element@0.0.7/demo/ananke/content/en/index.html) DOM which render same UI with common CSS in and out of shadow.

**Comment by sashafirsov** [comment]
> 2. Option A (no Shadow DOM) w/ caveats - no <slot>, 3rd party shadow boundaries

I do not see the need to exclude slots and templates. Those are essential part of WC and should be presented in light version as well. There is a [reference implementation](https://github.com/sashafirsov/light-dom-element) and [tooling lib](https://www.npmjs.com/package/css-chain#light-dom) to support templates in light DOM.

**Comment by thescientist13** [comment]
Yeah, in my mind I don't exclusively map Shadow DOM to Web Components, but I understand how compelling it is a feature to deliver a single file component experience, as it were.  Either way, I am hoping to understand the philosophical angle of this group as a whole, if it's worth making the push for WC in a light Dom world, or if it should just focus on making it as easy and clear to use WCs with Shadow DOM.  (Maybe this is what `adoptedStyleSheet` will solve, and so maybe a better way to focus our efforts instead of open stylable shadow roots?)
